lib/bridge/MemberListSyncer.js

```diff
@@ -247,10 +267,86 @@ function joinMatrixUsersToChannels(rooms, server, injectJoinFn) {
+        `rooms for server ${server.domain}`
+    );
+
+    // Store the matrix room info in memory for later retrieval when NAMES is received
+    // and updateIrcMemberList is called. At that point, we have enough information to
+    // leave users from the channel that the NAMES is for.
+    rooms.forEach((roomInfo) => {
+        this._memberLists.matrix[roomInfo.id] = roomInfo;
+    });
 }
 
+// Update the MemberListSyncer with the IRC NAMES_RPL that has been received for channel.
+// This will leave any matrix users that do not have their associated IRC nick in the list
+// of names for this channel.
+MemberListSyncer.prototype.updateIrcMemberList = Promise.coroutine(function*(channel, names) {
+    log.info(`updateIrcMemberList: Updating IRC member list for ${channel}`);
+    if (this._memberLists.irc[channel] !== undefined) {
+        log.info(`Names already received for ${channel}`);
+        return;
+    }
+    if (!this.server.shouldSyncMembershipToMatrix("initial", channel)) {
+        log.info(`Initial membership list syncing disabled for ${channel} and/or globally`);
+        return;
+    }
+
+    this._memberLists.irc[channel] = Object.keys(names);
+
+    // Convert the IRC channels nicks to userIds
+    let ircUserIds = this._memberLists.irc[channel].map(
+        (nick) => this.server.getUserIdFromNick(nick)
+    );
+
+    // For all bridged rooms, leave users from matrix that are not in the channel
+    let roomsForChannel = yield this.ircBridge.getStore().getMatrixRoomsForChannel(
+        this.server, channel
+    );
+
+    if (roomsForChannel.length === 0) {
+        log.info(`updateIrcMemberList: No bridged rooms for channel ${channel}`);
+        return;
+    }
+
+    let bridgedRoomIds = roomsForChannel.map((matrixRoom) => {
+        return matrixRoom.getId();
+    }).filter((roomId) => { // Filter rooms to remove those without member list data stored
+        return this._memberLists.matrix[roomId] &&
+            this._memberLists.matrix[roomId].remoteJoinedUsers;
+    });
+
+    if (bridgedRoomIds.length === 0) {
+        log.info(
+            `updateIrcMemberList: No rooms with known membership list data ` +
+            `bridged to channel ${channel}`
+        );
+        return;
+    }
+
+    // If a userId is in remoteJoinedUsers, but not ircUserIds, intend on leaving roomId
+    let promises = bridgedRoomIds.map((roomId) => {
+        let userIdsToLeave = this._memberLists.matrix[roomId].remoteJoinedUsers.filter(
+            (userId) => {
+                return ircUserIds.indexOf(userId) === -1;
+            }
+        );
+        log.info(
+            `updateIrcMemberList: Leaving ${userIdsToLeave.length} users from ${roomId} ` +
+            `because they are not in ${channel}.`
+        );
+        return userIdsToLeave.map(
+            (userId) => {
+                return this.ircBridge.getAppServiceBridge().getIntent(userId).leave(roomId);
+            }
+        );
+    }).reduce((a, b) => a.concat(b));
```

Comment:
Still feels too complicated to me: you have a map -> filter -> map -> reduce -> concat, whereas what you're really just doing is:
```js
let promises = [];
bridgedRoomIds.forEach((roomId) => {
    this._memberLists.matrix[roomId].remoteJoinedUsers.forEach((userId) => {
        if (ircUserIds.indexOf(userId) === -1) {
            promises.push(this.ircBridge.getAppServiceBridge().getIntent(userId).leave(roomId));
        }
    });
});
log.info(
    `updateIrcMemberList: Leaving ${userIdsToLeave.length} users from ${roomId} ` +
    `because they are not in ${channel}.`
);
yield Promise.all(promises);
```

Not only is this about the same lines of code, it's more performant (fewer loops) and IMHO clearer to see what you're manipulating.